includes/compat/subscriptions/class-wc-payment-gateway-wcpay-subscriptions-compat.php

```diff
@@ -37,5 +39,71 @@ public function __construct( ...$args ) {
 				'multiple_subscriptions',
 			]
 		);
+
+		add_action( 'woocommerce_scheduled_subscription_payment_' . $this->id, [ $this, 'scheduled_subscription_payment' ], 10, 2 );
+		add_filter( 'wcs_new_order_created', [ $this, 'copy_token_from_parent_order' ], 10, 2 );
+		add_action( 'woocommerce_subscription_failing_payment_method_updated_' . $this->id, [ $this, 'update_failing_payment_method' ], 10, 2 );
+		add_filter( 'wc_payments_display_save_payment_method_checkbox', [ $this, 'display_save_payment_method_checkbox' ], 10 );
+	}
+
+	public function process_payment( $order_id, $is_recurring_payment = false ) {
+		return parent::process_payment( $order_id, wcs_order_contains_subscription( $order_id ) );
+	}
+
+	public function display_save_payment_method_checkbox( $display ) {
+		if ( WC_Subscriptions_Cart::cart_contains_subscription() ) {
+			return false;
+		}
+		// Only render the "Save payment method" checkbox if there are no subscription products in the cart.
+		return $display;
+	}
+
+	/**
+	 * Scheduled_subscription_payment function.
+	 *
+	 * @param $amount float The amount to charge.
+	 * @param $renewal_order WC_Order A WC_Order object created to record the renewal payment.
+	 */
+	public function scheduled_subscription_payment( $amount, $renewal_order ) {
+		$order_tokens = WC_Payment_Tokens::get_order_tokens( $renewal_order );
+		$token        = empty( $order_tokens ) ? false : end( $order_tokens );
+		if ( ! $token ) {
+			Logger::error( 'There is no saved payment token for order #' . $renewal_order->get_id() );
+			$renewal_order->update_status( 'failed' );
+			return;
+		}
+
+		try {
+			$this->process_payment_for_order(
+				[
+					'order'           => $renewal_order,
+					'cart'            => null,
+					'manual_capture'  => false,
+					'payment_method'  => $token->get_token(),
+					'token'           => $token,
+					'is_saved_method' => true,
+					'off_session'     => true,
+					'is_recurring'    => true,
+				]
+			);
+		} catch ( WC_Payments_API_Exception $e ) {
+			Logger::error( 'Error processing subscription renewal: ' . $e->getMessage() );
+
+			$renewal_order->update_status( 'failed' );
+		}
+	}
+
+	public function copy_token_from_parent_order( $order, $subscription ) {
+		$parent = $subscription->get_parent();
```

Comment:
Relying on the tokens from the parent order can lead to incorrect cards being used to process recurring payments. 

Here's an example of how that can happen:

1. Customer checks out using card ending in 4242, a token for that card is saved on the order created at checkout and the associated subscription/s
1. First renewal payment is processed and uses card 4242 as expected
1. Customer changes card on subscription to new card ending in 4141
1. Next renewal payment pulls card from parent order, meaning it finds the token used to sign-up with card 4242 instead of the card 4141 now set as the payment method on the subscription

The good news is that tokens should be on the subscription after checkout without any effort from WC Pay (because `wcs_copy_order_meta()` should have copied over the `'_payment_tokens'` meta from the parent order to the subscription).

If that's the case, nothing extra should be needed to set the tokens on the new order, and `copy_token_from_parent_order()` can be deleted, because the renewal `$order` will already have the same `'_payment_tokens'` after they were copied over from the subscription in `wcs_create_renewal_order()` / `wcs_create_order_from_subscription()`). 

If for some reason the tokens do need to be copied manually, then you could avoid the incorrect card/token issue by using `$subscription->get_payment_tokens()` here instead of `$parent->get_payment_tokens()`.

Reply:
Great catch! This is actually something that was messing with the payment method change since the subscription order had the payment method altered, but the parent did not. I've removed this code as part of #832 in https://github.com/Automattic/woocommerce-payments/pull/832/commits/b313efb7925740bdbd8856966831aad88ccce255.

Reply:
After I removed the code to copy the payment token from the parent order, subscriptions were left without any payment token and I had to add https://github.com/Automattic/woocommerce-payments/pull/832/commits/efe43d76248da362efb884c51a5412ed0bf4ef0c.

Are we expected to set the metadata for subscriptions when processing the parent order, or should it be copied automatically (similar to renewals)?

Reply:
@luizreis **Short answer**: if subscriptions were left without any payment token, then SHA: efe43d7 is required and a good approach.

**Long answer**: whether the tokens are copied to the subscription should depend only on when the token meta data is being set on the order. If the token is set on the order _after_ the subscription is created, the meta will need to also be set on the subscription manually (normally, the subscription is created before payment is processed though). If its set before, then it should be copied automatically, but I might be missing something that prevents that happening, like the meta data being filtered out on the `'wcs_subscription_meta'` filter.

Regardless, Subscriptions is very unopinionated about tokens (for better or worse), so WC Pay and other gateways can do whatever works best.

Reply:
@luizreis Since you already seem to have handled this in #832 should we keep the changes there, or are there some additional changes I should make in this PR first?

Reply:
@thenbrent - gotcha! Thank you for the detailed answer. Since we're only adding the token once the `setup_intent` is successful, the metadata is not in the parent order before `process_payment` is called and that's the reason the subscription order doesn't have it.

@reykjalin - good question. There are no changes required in this PR, but I think it might benefit from `cherry-picking` https://github.com/Automattic/woocommerce-payments/pull/832/commits/87d39ef87421158f158b309a1e827142776cdf2d and https://github.com/Automattic/woocommerce-payments/pull/832/commits/249765c1f40d8bce9e77f313532764bfbd6be4ff so we don't have to review code that's "already" removed. What do you think?

Reply:
I've cherry-picked the 2 commits in https://github.com/Automattic/woocommerce-payments/pull/826/commits/0fe09c2c9e0c4e1eb976f8b2acabdd06c3be5f43 and https://github.com/Automattic/woocommerce-payments/pull/826/commits/1b3692bd7975dea897c462178c2ebc7694a0e94e 🎉

Reply:
Does that mean we can mark this as resolved, or is there further action we need to take in this PR?

Reply:
Yes, we can. There's a small fix that I'll need to apply on #832, but it's more related to #698 than this one. Thank you for adding those commits here!